### PR TITLE
Python: Multiple django improvements

### PR DESCRIPTION
**Pull request by RasmusWL**
This PR covers 3 improvements to our django modeling, so we now handle:

- view functions that have decorators
- classes used with `.as_view()` in routing setup, but without known view super-type
- taint of `self.request`, `self.args`, and `self.kwargs` on view class

I can be persuaded to split this into 3 PRs, I just found it easy to do them together. Will strongly recommend going through changes commit-by-commit

### CODE REVIEW

File: python/ql/src/semmle/python/frameworks/Django.qll [hunk 1]
```diff
@@ -1987,7 +1995,18 @@ private module Django {
...
+    (
+      not exists(func.getADecorator()) and
+      result.asExpr() = func.getDefinition()
+      or
+      // If the function has decorators, we still want to model the function as being
+      // the request handler for a route setup. In such situations, we must track the
+      // last decorator call instead of the function itself.
+      //
+      // Note that this means that we blindly ignore what the decorator actually does to
+      // the function, which seems like an OK tradeoff.
```
Comment (yoff) [inline python/ql/src/semmle/python/frameworks/Django.qll:2007]: Should we at least check that the decorator has anything to do with request handling (and possibly Django)?
We could for instance have a list of function names to compare with.

File: python/ql/src/semmle/python/frameworks/Django.qll [hunk 2]
```diff
@@ -2029,7 +2029,7 @@ private module Django {
     DjangoViewClassDef() { this.getABase() = django::views::generic::View::subclassRef().asExpr() }
 
     /** Gets a function that could handle incoming requests, if any. */
-    DjangoRouteHandler getARequestHandler() {
+    Function getARequestHandler() {
```
Comment (yoff) [inline python/ql/src/semmle/python/frameworks/Django.qll]: Is this change related/necessary?

File: python/ql/src/semmle/python/frameworks/Django.qll [hunk 3]
```diff
@@ -2078,7 +2099,7 @@ private module Django {
     DjangoRouteHandler() {
       exists(DjangoRouteSetup route | route.getViewArg() = djangoRouteHandlerFunctionTracker(this))
       or
-      any(DjangoViewClassDef vc).getARequestHandler() = this
+      any(DjangoViewClass vc).getARequestHandler() = this
```
Comment (yoff) [inline python/ql/src/semmle/python/frameworks/Django.qll:2125]: Should this (and below) use `PossibleDjangoViewClass`?

File: python/ql/src/semmle/python/frameworks/Django.qll [hunk 4]
```diff
@@ -2024,18 +2024,8 @@ private module Django {
...
-    /** Gets a function that could handle incoming requests, if any. */
-    Function getARequestHandler() {
-      // TODO: This doesn't handle attribute assignment. Should be OK, but analysis is not as complete as with
-      // points-to and `.lookup`, which would handle `post = my_post_handler` inside class def
-      result = this.getAMethod() and
-      result.getName() = HTTP::httpVerbLower()
-    }
-
+  /** A class that might be a django View class. */
+  class PossibleDjangoViewClass extends Class {
```
Comment (yoff) [inline python/ql/src/semmle/python/frameworks/Django.qll:2028]: Is the name meant to make me think of the `Range` pattern?

**Review by yoff — COMMENTED** [review]
Overall looks like nice improvements (good catch that `djangoRouteHandlerFunctionTracker` still tracks _all_ functions).

File: python/ql/src/semmle/python/frameworks/Django.qll (hunk 1 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Django.qll:2007]: I would rather not. In the real-code base that motivated this change, it was a decorator from a very small third-party library that we would probably never invest time in manually modeling.

I'm optimistic about this change, and think it'll just work out. If that turns out to be wrong, I'm very happy to change my mind. 

However, I think allowing any decorator align with our current goals of being able to find more results, and accepting that there might be extra FPs,

File: python/ql/src/semmle/python/frameworks/Django.qll (hunk 2 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Django.qll]: This change was necessary to get `DjangoRouteHandler` into a good state. Without this change, I saw that _no_ request handlers on classes was included in the `DjangoRouteHandler` class. I assume this is because the `getARequestHandler` predicate now will only give you results that are already considered a `DjangoRouteHandler`, so essentially we can remove the second part of the char-pred (since it doesn't specify any new functions that should be included in the `DjangoRouteHandler` class)

```diff
DjangoRouteHandler() {
 exists(DjangoRouteSetup route | route.getViewArg() = djangoRouteHandlerFunctionTracker(this))
-or
-any(DjangoViewClass vc).getARequestHandler() = this
}
```

File: python/ql/src/semmle/python/frameworks/Django.qll (hunk 3 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Django.qll:2125]: Nope. `PossibleDjangoViewClass` might be a bit too strongly worded. Maybe it should have been called `ProbablyNotDjangoViewClass` :laughing:

File: python/ql/src/semmle/python/frameworks/Django.qll (hunk 4 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Django.qll:2028]: not really

File: python/ql/src/semmle/python/frameworks/Django.qll (hunk 1 above)
Comment (yoff) [reply python/ql/src/semmle/python/frameworks/Django.qll:2007]: Allright, ship to learn 🚀

File: python/ql/src/semmle/python/frameworks/Django.qll (hunk 2 above)
Comment (yoff) [reply python/ql/src/semmle/python/frameworks/Django.qll]: Ah, so `getARequestHandler` is the more basic predicate.

**Review by yoff — COMMENTED** [review]
Thanks for the explanations. I am still not quite sure of the role of/need for `PossibleDjangoViewClass`. It seems like a not-meant-to-use-further extension point.
Apart from this, I am happy to merge this.

**Comment by RasmusWL** [comment]
The role of `PossibleDjangoViewClass` is to enable us to do this bit _Improved modeling of django View classes. We now consider any class using in a routing setup with `<class>.as_view()` as django view class._ 

`PossibleDjangoViewClass` is _not_ an extension point (and it is not abstract).

Since we no longer want to restrict the classes considered for tracking `<class>.as_view()`, We need a to include two sets of classes, to be considered a django view class (possibly with overlap):

- the classes that have a known view class as a super-class
- the classes that are used with `<class>.as_view()`

The sole purpose of `PossibleDjangoViewClass` is to provide tracking mechanisms for `<class>.as_view()` for all classes. Since it's all classes, this models all possible future django view classes.

Hopefully this makes things more clear. Otherwise let's talk on zoom. If you do get an _aha_ moment, would probably be cool with a docs suggestion so it's more clear for the next person :shrug:

**Review by yoff — APPROVED** [review]
Thanks, LGTM
